## Notes from NumPy Data Science Essential Training

### 1. Create NumPy arrays using Python's "array like" data types

In [1]:
# Import numpy package and check its version
import numpy as np
print(np.__version__)

1.16.4


#### Array from a list

In [2]:
my_list = [-17, 0, 4, 5, 9]
my_array_from_list = np.array(my_list)
my_array_from_list

array([-17,   0,   4,   5,   9])

#### Array from a tuple

In [3]:
my_tuple = (14, -3.54, 5+7j)
np.array(my_tuple)

array([14.  +0.j, -3.54+0.j,  5.  +7.j])

#### Difference between tuples and numpy arrays

In [6]:
# Multiplying a tuple by a scalar "n" duplicate the tuple n times
my_tuple * 5

(14,
 -3.54,
 (5+7j),
 14,
 -3.54,
 (5+7j),
 14,
 -3.54,
 (5+7j),
 14,
 -3.54,
 (5+7j),
 14,
 -3.54,
 (5+7j))

In [7]:
# Multiplying an array by a scalar "n", produce an array whose elements are the elements 
# of the original array multiplied by n 
np.array(my_tuple) * 5

array([ 70.  +0.j, -17.7 +0.j,  25. +35.j])

#### Create an array using np.arrange()

In [8]:
# The most basic way
np.arange(7)

array([0, 1, 2, 3, 4, 5, 6])

In [9]:
# Array whose elements are comprised bewteen two numbers
np.arange(10, 23)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22])

In [10]:
# With a step of 5
np.arange(0, 26, step=5)
# Or simply: np.arange(0, 26, 5)

array([ 0,  5, 10, 15, 20, 25])

#### Size of an array

In [11]:
np.arange(10, 23).size

13

In [12]:
# this is the same as using len()
len(np.arange(10, 23))

13

#### Create an array using linspace()

In [13]:
import numpy as np
np.linspace(5, 15, 9)

array([ 5.  ,  6.25,  7.5 ,  8.75, 10.  , 11.25, 12.5 , 13.75, 15.  ])

In [14]:
# To find out the step size used
my_linspace = np.linspace(5, 15, 9, retstep=True)
my_linspace[1]

1.25

In [17]:
# To display the array 
my_linspace[0]

array([ 5.  ,  6.25,  7.5 ,  8.75, 10.  , 11.25, 12.5 , 13.75, 15.  ])

#### Create an array using zeros()

In [20]:
import numpy as np
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [22]:
# Array of dim 3 * 5 * 4
np.zeros((3,5,4))

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])

#### Create an array using ones()

In [21]:
import numpy as np
np.ones(7)

array([1., 1., 1., 1., 1., 1., 1.])

In [24]:
# Array of dim 4 * 5
np.ones((4,5))

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

#### Create an array with a predefined data type

In [25]:
# Array of integers
np.zeros(11, dtype='int64')

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [26]:
# Array of floats
np.zeros(11)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

### 2. Array's indexing

#### For vectors

In [27]:
import numpy as np

# First we create an array to use
my_vector = np.array ([-17, -4, 0, 2, 21, 37, 105])
my_vector

array([-17,  -4,   0,   2,  21,  37, 105])

In [28]:
# Get the first element
my_vector[0]

-17

In [29]:
# Change the first element
my_vector[0] = -102
my_vector

array([-102,   -4,    0,    2,   21,   37,  105])

In [30]:
# Display the 3rd to last element
my_vector[-3]

21

In [31]:
# Indexing after an operation
my_vector[305 % 7]

21

#### For two-dimensional arrays

In [32]:
# Create a two-dimensional array to use
# First we create a vector of 35 elements
my_array= np.arange(35)
# Second we reshape it into a 7 by 5 array
my_array.shape = (7,5)
my_array

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34]])

In [33]:
# Display the second row
my_array[2]

array([10, 11, 12, 13, 14])

In [37]:
# Display the second to last column
my_array[:,-2]

array([ 3,  8, 13, 18, 23, 28, 33])

In [38]:
# Display a specific element of an array
my_array[5,2]

27

#### For three-dimensional arrays

In [40]:
# Create a three-dimensional array to use
# First we create a vector of 35 elements
my_3D_array= np.arange(70)
# Second we reshape it into a 2 by 7 by 5 array
my_3D_array.shape = (2, 7, 5)
my_3D_array

array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34]],

       [[35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44],
        [45, 46, 47, 48, 49],
        [50, 51, 52, 53, 54],
        [55, 56, 57, 58, 59],
        [60, 61, 62, 63, 64],
        [65, 66, 67, 68, 69]]])

In [41]:
# Display the elements in the second position of the first dimension
my_3D_array[1]

array([[35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44],
       [45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54],
       [55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64],
       [65, 66, 67, 68, 69]])

In [43]:
# Display the elements that are in the 4th row of the second two dimensional array 
my_3D_array[1,3]

array([50, 51, 52, 53, 54])

In [45]:
# Change a specific element in the array
my_3D_array[1,3,2] = 1111
my_3D_array

array([[[   0,    1,    2,    3,    4],
        [   5,    6,    7,    8,    9],
        [  10,   11,   12,   13,   14],
        [  15,   16,   17,   18,   19],
        [  20,   21,   22,   23,   24],
        [  25,   26,   27,   28,   29],
        [  30,   31,   32,   33,   34]],

       [[  35,   36,   37,   38,   39],
        [  40,   41,   42,   43,   44],
        [  45,   46,   47,   48,   49],
        [  50,   51, 1111,   53,   54],
        [  55,   56,   57,   58,   59],
        [  60,   61,   62,   63,   64],
        [  65,   66,   67,   68,   69]]])

### 3. Boolean Mask Arrays

In [1]:
import numpy as np

# First we create an array to use
my_vector = np.array([-17, -4, 0, 2, 21, 37, 105])
my_vector

array([-17,  -4,   0,   2,  21,  37, 105])

In [2]:
# Check whether each element of the vector is divisible by 7 or not
zero_mod_7_mask = (0 == (my_vector % 7))
zero_mod_7_mask

array([False, False,  True, False,  True, False,  True])

#### Select the elements that are divisible by 7 and are positive: method 1

In [3]:
# Select the elements that are divisible by 7
sub_array_div7 = my_vector[zero_mod_7_mask]
sub_array_div7

array([  0,  21, 105])

In [5]:
# Select the elements that are divisible by 7 and are positive
sub_array_div7[sub_array_div7>0]

array([ 21, 105])

#### Select the elements that are divisible by 7 and are positive: method 2

In [6]:
# Test which elements are divisible by 7
mod_test = (0 == (my_vector % 7))
mod_test

array([False, False,  True, False,  True, False,  True])

In [7]:
# Test which elements are > 0
positive_test = my_vector > 0
positive_test

array([False, False, False,  True,  True,  True,  True])

In [8]:
# Test which elements are divisible by 7 and are > 0
combined_mask = np.logical_and(mod_test, positive_test)
combined_mask

array([False, False, False, False,  True, False,  True])

In [9]:
# Select the elements that are divisible by 7 and are positive
my_vector[combined_mask]

array([ 21, 105])

In [17]:
# logical_and can also work with multiple arrays
np.logical_and(np.array([True, False]), np.array([True, True]), np.array([True, True]))

array([ True, False])

### 4. Operations with arrays

#### Exploring some useful functions 

In [26]:
import numpy as np

# Define an array to work with
my_3D_array = np.arange(70)
my_3D_array.shape = (2,7,5)
my_3D_array

array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34]],

       [[35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44],
        [45, 46, 47, 48, 49],
        [50, 51, 52, 53, 54],
        [55, 56, 57, 58, 59],
        [60, 61, 62, 63, 64],
        [65, 66, 67, 68, 69]]])

In [19]:
# shape
my_3D_array.shape

(2, 7, 5)

In [20]:
# number of dimensions
my_3D_array.ndim

3

In [21]:
# size, i.e., number of elements
my_3D_array.size

70

In [22]:
# data type for each element
my_3D_array.dtype

dtype('int32')

In [27]:
# item size for each element within an array
my_3D_array.itemsize

4

In [28]:
# type (for the array; not for elements within array)
type(my_3D_array)

numpy.ndarray

#### Operations involving scalars

In [23]:
5 * my_3D_array - 2

array([[[ -2,   3,   8,  13,  18],
        [ 23,  28,  33,  38,  43],
        [ 48,  53,  58,  63,  68],
        [ 73,  78,  83,  88,  93],
        [ 98, 103, 108, 113, 118],
        [123, 128, 133, 138, 143],
        [148, 153, 158, 163, 168]],

       [[173, 178, 183, 188, 193],
        [198, 203, 208, 213, 218],
        [223, 228, 233, 238, 243],
        [248, 253, 258, 263, 268],
        [273, 278, 283, 288, 293],
        [298, 303, 308, 313, 318],
        [323, 328, 333, 338, 343]]])

#### Array multiplication

In [25]:
# Create two arrays to use
left_mat = np.arange(6).reshape((2,3))
right_mat = np.arange(15).reshape((3,5))
# dot product
np.dot(left_mat, right_mat)

array([[ 25,  28,  31,  34,  37],
       [ 70,  82,  94, 106, 118]])

#### Inner product

In [31]:
# Create two vectors to use
left_vect = np.arange(5)
right_vect = np.arange(7,12)
np.inner(left_vect, right_vect)

100

#### sum along axes 

In [32]:
# Array to use
my_3D_array

array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34]],

       [[35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44],
        [45, 46, 47, 48, 49],
        [50, 51, 52, 53, 54],
        [55, 56, 57, 58, 59],
        [60, 61, 62, 63, 64],
        [65, 66, 67, 68, 69]]])

In [33]:
# Sum all elements of an array
my_3D_array.sum()

2415

In [34]:
# Sum along the first axis
my_3D_array.sum(axis=0)

array([[ 35,  37,  39,  41,  43],
       [ 45,  47,  49,  51,  53],
       [ 55,  57,  59,  61,  63],
       [ 65,  67,  69,  71,  73],
       [ 75,  77,  79,  81,  83],
       [ 85,  87,  89,  91,  93],
       [ 95,  97,  99, 101, 103]])

In [35]:
# Sum along the second axis
my_3D_array.sum(axis=1)

array([[105, 112, 119, 126, 133],
       [350, 357, 364, 371, 378]])

In [37]:
# Sum along the third/last axis
my_3D_array.sum(axis=2)

array([[ 10,  35,  60,  85, 110, 135, 160],
       [185, 210, 235, 260, 285, 310, 335]])

#### Multiplying arrays of different dimensions

In [52]:
# Set the number of decimal digits to display
np.set_printoptions(precision=3)
# Create a 2D-array
my_random_2D_array = np.random.random((7,5))
my_random_2D_array

array([[0.154, 0.946, 0.176, 0.786, 0.506],
       [0.827, 0.375, 0.309, 0.362, 0.148],
       [0.054, 0.427, 0.629, 0.106, 0.695],
       [0.351, 0.99 , 0.922, 0.979, 0.083],
       [0.654, 0.725, 0.845, 0.572, 0.285],
       [0.976, 0.891, 0.139, 0.339, 0.447],
       [0.054, 0.006, 0.206, 0.311, 0.829]])

In [54]:
# Multiplying a 2D and 3D arrays (element by element multiplication)
print(f"- The shape of the 3D array: {my_3D_array.shape}")
print(f"- The shape of the 2D array: {my_random_2D_array.shape}")
product_array = my_3D_array * my_random_2D_array
print(f"- The shape of the product array: {product_array.shape}")
print(f"- The product array: \n {product_array}")

- The shape of the 3D array: (2, 7, 5)
- The shape of the 2D array: (7, 5)
- The shape of the product array: (2, 7, 5)
- The product array: 
 [[[ 0.     0.946  0.352  2.357  2.025]
  [ 4.137  2.249  2.166  2.897  1.331]
  [ 0.537  4.696  7.545  1.38   9.724]
  [ 5.265 15.845 15.675 17.621  1.574]
  [13.077 15.23  18.585 13.154  6.848]
  [24.412 23.177  3.753  9.498 12.963]
  [ 1.631  0.184  6.598 10.268 28.186]]

 [[ 5.381 34.072  6.514 29.85  19.74 ]
  [33.096 15.371 12.995 15.574  6.506]
  [ 2.416 19.636 29.55   5.096 34.035]
  [17.55  50.507 47.947 51.885  4.474]
  [35.96  40.612 48.152 33.171 16.835]
  [58.589 54.376  8.618 21.371 28.608]
  [ 3.534  0.393 13.815 21.159 57.201]]]


#### Dividing arrays of different dimensions

In [55]:
# Vector to use for the division
my_vector = np.arange(5) * 7
my_vector[0] = -1
my_vector

array([-1,  7, 14, 21, 28])

In [59]:
# Dividing a 3D arrays by a 1D array (element by element division)
np.set_printoptions(precision=2)
print(f"- The 3D array: \n {my_3D_array}")
print(f"- The 1D array: \n {my_vector}")
print(f"- The array resulting from division: \n {my_3D_array / my_vector}")
#my_3D_array / my_vector

- The 3D array: 
 [[[ 0  1  2  3  4]
  [ 5  6  7  8  9]
  [10 11 12 13 14]
  [15 16 17 18 19]
  [20 21 22 23 24]
  [25 26 27 28 29]
  [30 31 32 33 34]]

 [[35 36 37 38 39]
  [40 41 42 43 44]
  [45 46 47 48 49]
  [50 51 52 53 54]
  [55 56 57 58 59]
  [60 61 62 63 64]
  [65 66 67 68 69]]]
- The 1D array: 
 [-1  7 14 21 28]
- The array resulting from division: 
 [[[ -0.     0.14   0.14   0.14   0.14]
  [ -5.     0.86   0.5    0.38   0.32]
  [-10.     1.57   0.86   0.62   0.5 ]
  [-15.     2.29   1.21   0.86   0.68]
  [-20.     3.     1.57   1.1    0.86]
  [-25.     3.71   1.93   1.33   1.04]
  [-30.     4.43   2.29   1.57   1.21]]

 [[-35.     5.14   2.64   1.81   1.39]
  [-40.     5.86   3.     2.05   1.57]
  [-45.     6.57   3.36   2.29   1.75]
  [-50.     7.29   3.71   2.52   1.93]
  [-55.     8.     4.07   2.76   2.11]
  [-60.     8.71   4.43   3.     2.29]
  [-65.     9.43   4.79   3.24   2.46]]]


In [58]:
# Modulo operation
my_3D_array % my_vector

array([[[ 0,  1,  2,  3,  4],
        [ 0,  6,  7,  8,  9],
        [ 0,  4, 12, 13, 14],
        [ 0,  2,  3, 18, 19],
        [ 0,  0,  8,  2, 24],
        [ 0,  5, 13,  7,  1],
        [ 0,  3,  4, 12,  6]],

       [[ 0,  1,  9, 17, 11],
        [ 0,  6,  0,  1, 16],
        [ 0,  4,  5,  6, 21],
        [ 0,  2, 10, 11, 26],
        [ 0,  0,  1, 16,  3],
        [ 0,  5,  6,  0,  8],
        [ 0,  3, 11,  5, 13]]], dtype=int32)

### 5. Creating Structured Arrays

#### Uni-dimensional arrays

In [60]:
import numpy as np

# Define the structure of the array
# The 1st element is a string labelled name.
# The 2nd element is a floating number labelled height.
# The 3rd element is a floating number labelled weight.
# The 4th element is an integer number labelled age.
person_data_def = [('name','S6'),('height','f8'),('weight','f8'), ('age', 'i8')]
person_data_def

[('name', 'S6'), ('height', 'f8'), ('weight', 'f8'), ('age', 'i8')]

In [61]:
# Create an array that has the above structure
people_array = np.zeros((4,), dtype=person_data_def)
people_array

array([(b'', 0., 0., 0), (b'', 0., 0., 0), (b'', 0., 0., 0),
       (b'', 0., 0., 0)],
      dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [62]:
# Modify some of its elements
people_array[3] = ('Delta', 73, 205, 34)
people_array[0] = ('Alpha', 65, 112, 23)
people_array

array([(b'Alpha', 65., 112., 23), (b'',  0.,   0.,  0),
       (b'',  0.,   0.,  0), (b'Delta', 73., 205., 34)],
      dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [63]:
# Select/display some elements of the array
people_array[2:]

array([(b'',  0.,   0.,  0), (b'Delta', 73., 205., 34)],
      dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [64]:
# Extract all the available values of a specific attribute (here age)
ages = people_array['age']
ages

array([23,  0,  0, 34], dtype=int64)

#### Multi-dimensional structured arrays

In [66]:
# Create a multi-dimensional array that has the above structure
people_big_array = np.zeros((4,3,2), dtype=person_data_def)
people_big_array

array([[[(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)]],

       [[(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)]],

       [[(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)]],

       [[(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)],
        [(b'', 0., 0., 0), (b'', 0., 0., 0)]]],
      dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [67]:
# Modify some of its elements
people_big_array[3,2,1] = ('Echo', 68, 155, 46)
people_big_array

array([[[(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)]],

       [[(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)]],

       [[(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)]],

       [[(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'',  0.,   0.,  0)],
        [(b'',  0.,   0.,  0), (b'Echo', 68., 155., 46)]]],
      dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [71]:
# Extract all the available values of a specific attribute (here height)
people_big_array['height']

array([[[ 0.,  0.],
        [ 0.,  0.],
        [ 0.,  0.]],

       [[ 0.,  0.],
        [ 0.,  0.],
        [ 0.,  0.]],

       [[ 0.,  0.],
        [ 0.,  0.],
        [ 0.,  0.]],

       [[ 0.,  0.],
        [ 0.,  0.],
        [ 0., 68.]]])

In [70]:
# Extract all the available values of multiple attributes (here height and weight)
people_big_array[['height', 'weight']]

array([[[( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)]],

       [[( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)]],

       [[( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)]],

       [[( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), ( 0.,   0.)],
        [( 0.,   0.), (68., 155.)]]],
      dtype=[('height', '<f8'), ('weight', '<f8')])

### 6. Creating Record Arrays

Can make it easier to extract information based on attributes instead of indices as in structured arrays

In [72]:
# Define a record array
person_record_array = np.rec.array([('Delta', 73, 205, 34),('Alpha', 65, 112, 23)],dtype=person_data_def)
person_record_array

rec.array([(b'Delta', 73., 205., 34), (b'Alpha', 65., 112., 23)],
          dtype=[('name', 'S6'), ('height', '<f8'), ('weight', '<f8'), ('age', '<i8')])

In [74]:
# Extract the age of the first element in the array
person_record_array[0].age

34

### 7. Views and Copies

- A copy takes data from one location in memory and replicates it in a different location in memory.

- A view provides two differently named references to the same location in memory

#### Creating a copy of an array in the same memory location (a shallow copy)

In [1]:
import numpy as np

# Create an array to use
mi_casa = np.array([-45, -31, -12, 0, 2, 25, 51, 99])
mi_casa

array([-45, -31, -12,   0,   2,  25,  51,  99])

In [7]:
# Create an array equal to mi_casa
su_casa = mi_casa

# test reference equality
print(f"- Reference equality test: {mi_casa is su_casa}")

# test value equality
print(f"- Value equality test: {mi_casa == su_casa}")

- Reference equality test: True
- Value equality test: [ True  True  True  True  True  True  True  True]


In [5]:
# Memory location of the two arrays
print(f"- The location of mi_casa is: {id(mi_casa)}")
print(f"- The location of su_casa is: {id(su_casa)}")

- The location of mi_casa is: 2555961074032
- The location of su_casa is: 2555961074032


#### Another way to create a shallow copy

In [19]:
tree_house = np.array([-45, -31, -12, 0, 2, 25, 51, 99])
farm_house = tree_house.view()
farm_house.shape = (2,4)

In [20]:
print(f"- tree_house: {tree_house}")
print(f"- farm_house: {farm_house}")

- tree_house: [-45 -31 -12   0   2  25  51  99]
- farm_house: [[-45 -31 -12   0]
 [  2  25  51  99]]


In [21]:
# Changing one element in tree_house alternate the same element
# in farm house although they have different shapes
tree_house [3] = -111
print(f"- tree_house: {tree_house}")
print(f"- farm_house: {farm_house}")

- tree_house: [ -45  -31  -12 -111    2   25   51   99]
- farm_house: [[ -45  -31  -12 -111]
 [   2   25   51   99]]


#### Creating a deep copy

In [14]:
tree_house = np.array([-45, -31, -12, 0, 2, 25, 51, 99])
dog_house = np.copy(tree_house)

In [15]:
print(f"- tree_house: {tree_house}")
print(f"- dog_house: {dog_house}")

- tree_house: [-45 -31 -12   0   2  25  51  99]
- dog_house: [-45 -31 -12   0   2  25  51  99]


In [23]:
# Changing one element in dog_house DOESN'T alternate the same element
# in tree house
dog_house[0] = -121
print(f"- tree_house: {tree_house}")
print(f"- dog_house: {dog_house}")

- tree_house: [ -45  -31  -12 -111    2   25   51   99]
- dog_house: [-121  -31  -12    0    2   25   51   99]


### 8. Adding and Removing Elements from NumPy Arrays
- append
- horizontal stacking
- vertical stacking
- insert
- delete

#### Append an array to another array of a different shape

In [2]:
import numpy as np

# Create an array to start with
a = np.array(np.arange(24)).reshape(2,3,4)
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [3]:
# A vector is obtained if we append a vector to the multi-dimensional array 
b = np.append(a, [5,6,7,8])
b

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23,  5,  6,  7,  8])

#### Append on a specific axis

In [4]:
# First create an array of the same shape as "a"
c = np.array(np.arange(24)).reshape(2,3,4) * 10 + 3
print(f"- a is:\n {a}")
print(f"- c is:\n {c}")

- a is:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- c is:
 [[[  3  13  23  33]
  [ 43  53  63  73]
  [ 83  93 103 113]]

 [[123 133 143 153]
  [163 173 183 193]
  [203 213 223 233]]]


In [5]:
# Append on the first axis
a0 = np.append(a,c, axis=0)
a0

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23]],

       [[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [6]:
# Append on the second axis
a1 = np.append(a,c, axis=1)
a1

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11],
        [  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23],
        [123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [7]:
# Append on the third axis
a2 = np.append(a,c, axis=2)
a2

array([[[  0,   1,   2,   3,   3,  13,  23,  33],
        [  4,   5,   6,   7,  43,  53,  63,  73],
        [  8,   9,  10,  11,  83,  93, 103, 113]],

       [[ 12,  13,  14,  15, 123, 133, 143, 153],
        [ 16,  17,  18,  19, 163, 173, 183, 193],
        [ 20,  21,  22,  23, 203, 213, 223, 233]]])

#### horizontal stacking

In [8]:
# We will use the previously defined arrays a and c
print(f"- a is:\n {a}")
print(f"- c is:\n {c}")

- a is:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- c is:
 [[[  3  13  23  33]
  [ 43  53  63  73]
  [ 83  93 103 113]]

 [[123 133 143 153]
  [163 173 183 193]
  [203 213 223 233]]]


In [9]:
# Horizontaly stack a and c
my_hay_stack = np.hstack((a,c))
my_hay_stack

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11],
        [  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23],
        [123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

#### Inserting along a specific axis

In [10]:
# Insert along the first axis
after_insert_array = np.insert(c, 1, 444, axis=0)
after_insert_array

array([[[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[444, 444, 444, 444],
        [444, 444, 444, 444],
        [444, 444, 444, 444]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [11]:
# Insert along the second axis
np.insert(c, 1, 444, axis=1)

array([[[  3,  13,  23,  33],
        [444, 444, 444, 444],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [444, 444, 444, 444],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [12]:
# Insert along the third axis
np.insert(c, 1, 444, axis=2)

array([[[  3, 444,  13,  23,  33],
        [ 43, 444,  53,  63,  73],
        [ 83, 444,  93, 103, 113]],

       [[123, 444, 133, 143, 153],
        [163, 444, 173, 183, 193],
        [203, 444, 213, 223, 233]]])

#### Delete a vector (e.g., row or column) along a specific axis

In [13]:
# First we create an array to work with by copying c
d = np.empty(c.shape)
np.copyto(d, c)
d

array([[[   3.,   13.,   23.,   33.],
        [  43.,   53.,   63.,   73.],
        [  83.,   93.,  103.,  113.]],

       [[ 123.,  133.,  143.,  153.],
        [ 163.,  173.,  183.,  193.],
        [ 203.,  213.,  223.,  233.]]])

In [17]:
# Delete the second sub-array along the 1st dimension
e = np.delete(d, 1, axis=0)
e

array([[[   3.,   13.,   23.,   33.],
        [  43.,   53.,   63.,   73.],
        [  83.,   93.,  103.,  113.]]])

In [18]:
# Delete doesn't change the original element
d

array([[[   3.,   13.,   23.,   33.],
        [  43.,   53.,   63.,   73.],
        [  83.,   93.,  103.,  113.]],

       [[ 123.,  133.,  143.,  153.],
        [ 163.,  173.,  183.,  193.],
        [ 203.,  213.,  223.,  233.]]])

In [19]:
# Delete along the 2nd dimension
e = np.delete(d, 1, axis=1)
e

array([[[   3.,   13.,   23.,   33.],
        [  83.,   93.,  103.,  113.]],

       [[ 123.,  133.,  143.,  153.],
        [ 203.,  213.,  223.,  233.]]])

In [20]:
# Delete along the 3rd dimension
e = np.delete(d, 1, axis=2)
e

array([[[   3.,   23.,   33.],
        [  43.,   63.,   73.],
        [  83.,  103.,  113.]],

       [[ 123.,  143.,  153.],
        [ 163.,  183.,  193.],
        [ 203.,  223.,  233.]]])

### 9. Joining and Splitting arrays
- concatenate
- stack
- split

#### Concatenate along the first axis
documentation: http://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html

In [21]:
import numpy as np

# Create arrays to use
a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6]])
print(f"- a is:\n {a}")
print(f"- b is:\n {b}")

- a is:
 [[1 2]
 [3 4]]
- b is:
 [[5 6]]


In [22]:
# Concatenate a and b along the first axis
together = np.concatenate((a, b), axis=0)
together

array([[1, 2],
       [3, 4],
       [5, 6]])

#### Concatenate along the second axis

In [23]:
# Create a new array
c = np.array([[1, 2], [3, 4]]) *3 + 5
c

array([[ 8, 11],
       [14, 17]])

In [24]:
# Concatenate a and b along the second axis
np.concatenate((a, c), axis=1)

array([[ 1,  2,  8, 11],
       [ 3,  4, 14, 17]])

#### Stack along a certain axis

Join a sequence of arrays along a new axis.

In [27]:
# Create a three-dimensional array
arrays = np.zeros((5,3,4))
for n in range(5):
    arrays[n] = np.random.randn(3, 4)  
arrays

array([[[ -1.18757848e+00,  -7.91187933e-01,   7.02430216e-01,
           1.99311768e+00],
        [ -7.79279553e-01,  -8.68321983e-01,   9.84969259e-01,
          -1.50330293e+00],
        [  9.70289813e-01,   1.28465103e+00,   1.56144177e+00,
          -1.79158023e-01]],

       [[  1.06989763e+00,  -4.36415783e-01,   2.28594800e-01,
          -5.94739925e-01],
        [  1.14627666e+00,   5.24971640e-01,   5.97482241e-01,
          -1.15890587e+00],
        [  4.91227783e-01,   8.64914780e-01,   2.75329919e-01,
          -1.02358379e+00]],

       [[ -1.55908750e-01,   8.00559122e-01,   5.29809717e-01,
           5.99034948e-01],
        [  6.02226333e-01,  -3.06664144e-01,   5.05991975e-01,
           1.06283019e+00],
        [ -3.87990567e-01,  -2.50742057e-01,   1.84735730e-03,
           6.54435698e-01]],

       [[ -1.49093206e+00,   9.52493841e-01,   2.13022915e-02,
           2.02072982e+00],
        [  1.29165081e+00,   8.70727890e-01,  -2.19943252e+00,
           1.08072701

In [40]:
# Create three arrays by stacking "arrays" along each of the three axis 
stack0 = np.stack(arrays, axis=0)
stack1 = np.stack(arrays, axis=1)
stack2 = np.stack(arrays, axis=2)

# Check the shapes of the arrays
print(f"- Shape of arrays: {arrays.shape}\n"
      f"- Shape of stack0: {stack0.shape}\n" 
      f"- Shape of stack1: {stack1.shape}\n" 
      f"- Shape of stack2: {stack2.shape}")


- Shape of arrays: (5, 3, 4)
- Shape of stack0: (5, 3, 4)
- Shape of stack1: (3, 5, 4)
- Shape of stack2: (3, 4, 5)


In [47]:
# Display stack0
print(f"- original array:\n {arrays}")
print("\n")
print(f"- Stacked array along the first axis:\n {stack0}")

- original array:
 [[[ -1.18757848e+00  -7.91187933e-01   7.02430216e-01   1.99311768e+00]
  [ -7.79279553e-01  -8.68321983e-01   9.84969259e-01  -1.50330293e+00]
  [  9.70289813e-01   1.28465103e+00   1.56144177e+00  -1.79158023e-01]]

 [[  1.06989763e+00  -4.36415783e-01   2.28594800e-01  -5.94739925e-01]
  [  1.14627666e+00   5.24971640e-01   5.97482241e-01  -1.15890587e+00]
  [  4.91227783e-01   8.64914780e-01   2.75329919e-01  -1.02358379e+00]]

 [[ -1.55908750e-01   8.00559122e-01   5.29809717e-01   5.99034948e-01]
  [  6.02226333e-01  -3.06664144e-01   5.05991975e-01   1.06283019e+00]
  [ -3.87990567e-01  -2.50742057e-01   1.84735730e-03   6.54435698e-01]]

 [[ -1.49093206e+00   9.52493841e-01   2.13022915e-02   2.02072982e+00]
  [  1.29165081e+00   8.70727890e-01  -2.19943252e+00   1.08072701e-01]
  [ -2.28571420e-01   1.29973931e-01   8.23276176e-01   8.92589437e-02]]

 [[ -7.66375314e-02  -6.59462112e-01  -6.96985895e-01  -1.75645967e+00]
  [ -2.29347154e-01  -2.44364560e-01 

In [45]:
# Display stack1
print(f"- original array:\n {arrays}")
print("\n")
print(f"- Stacked array along the second axis:\n {stack1}")

- original array:
 [[[ -1.18757848e+00  -7.91187933e-01   7.02430216e-01   1.99311768e+00]
  [ -7.79279553e-01  -8.68321983e-01   9.84969259e-01  -1.50330293e+00]
  [  9.70289813e-01   1.28465103e+00   1.56144177e+00  -1.79158023e-01]]

 [[  1.06989763e+00  -4.36415783e-01   2.28594800e-01  -5.94739925e-01]
  [  1.14627666e+00   5.24971640e-01   5.97482241e-01  -1.15890587e+00]
  [  4.91227783e-01   8.64914780e-01   2.75329919e-01  -1.02358379e+00]]

 [[ -1.55908750e-01   8.00559122e-01   5.29809717e-01   5.99034948e-01]
  [  6.02226333e-01  -3.06664144e-01   5.05991975e-01   1.06283019e+00]
  [ -3.87990567e-01  -2.50742057e-01   1.84735730e-03   6.54435698e-01]]

 [[ -1.49093206e+00   9.52493841e-01   2.13022915e-02   2.02072982e+00]
  [  1.29165081e+00   8.70727890e-01  -2.19943252e+00   1.08072701e-01]
  [ -2.28571420e-01   1.29973931e-01   8.23276176e-01   8.92589437e-02]]

 [[ -7.66375314e-02  -6.59462112e-01  -6.96985895e-01  -1.75645967e+00]
  [ -2.29347154e-01  -2.44364560e-01 

In [46]:
# Display stack2
print(f"- original array:\n {arrays}")
print("\n")
print(f"- Stacked array along the third axis:\n {stack2}")

- original array:
 [[[ -1.18757848e+00  -7.91187933e-01   7.02430216e-01   1.99311768e+00]
  [ -7.79279553e-01  -8.68321983e-01   9.84969259e-01  -1.50330293e+00]
  [  9.70289813e-01   1.28465103e+00   1.56144177e+00  -1.79158023e-01]]

 [[  1.06989763e+00  -4.36415783e-01   2.28594800e-01  -5.94739925e-01]
  [  1.14627666e+00   5.24971640e-01   5.97482241e-01  -1.15890587e+00]
  [  4.91227783e-01   8.64914780e-01   2.75329919e-01  -1.02358379e+00]]

 [[ -1.55908750e-01   8.00559122e-01   5.29809717e-01   5.99034948e-01]
  [  6.02226333e-01  -3.06664144e-01   5.05991975e-01   1.06283019e+00]
  [ -3.87990567e-01  -2.50742057e-01   1.84735730e-03   6.54435698e-01]]

 [[ -1.49093206e+00   9.52493841e-01   2.13022915e-02   2.02072982e+00]
  [  1.29165081e+00   8.70727890e-01  -2.19943252e+00   1.08072701e-01]
  [ -2.28571420e-01   1.29973931e-01   8.23276176e-01   8.92589437e-02]]

 [[ -7.66375314e-02  -6.59462112e-01  -6.96985895e-01  -1.75645967e+00]
  [ -2.29347154e-01  -2.44364560e-01 

#### Split
Split an array into multiple sub-arrays. For information, please see: http://docs.scipy.org/doc/numpy/reference/generated/numpy.split.html

In [60]:
# Array to split
before_split = stack0
print(f"- Shape of the array to split: {before_split.shape}")
print(f"- Array to split:\n {before_split}")

- Shape of the array to split: (5, 3, 4)
- Array to split:
 [[[ -1.18757848e+00  -7.91187933e-01   7.02430216e-01   1.99311768e+00]
  [ -7.79279553e-01  -8.68321983e-01   9.84969259e-01  -1.50330293e+00]
  [  9.70289813e-01   1.28465103e+00   1.56144177e+00  -1.79158023e-01]]

 [[  1.06989763e+00  -4.36415783e-01   2.28594800e-01  -5.94739925e-01]
  [  1.14627666e+00   5.24971640e-01   5.97482241e-01  -1.15890587e+00]
  [  4.91227783e-01   8.64914780e-01   2.75329919e-01  -1.02358379e+00]]

 [[ -1.55908750e-01   8.00559122e-01   5.29809717e-01   5.99034948e-01]
  [  6.02226333e-01  -3.06664144e-01   5.05991975e-01   1.06283019e+00]
  [ -3.87990567e-01  -2.50742057e-01   1.84735730e-03   6.54435698e-01]]

 [[ -1.49093206e+00   9.52493841e-01   2.13022915e-02   2.02072982e+00]
  [  1.29165081e+00   8.70727890e-01  -2.19943252e+00   1.08072701e-01]
  [ -2.28571420e-01   1.29973931e-01   8.23276176e-01   8.92589437e-02]]

 [[ -7.66375314e-02  -6.59462112e-01  -6.96985895e-01  -1.75645967e+

In [56]:
# Split the array along the first axis
s0 = np.split(before_split, 5, axis=0)
print(f"- Type of the split array: {type(s0)}")
print(f"- Shape of the split array: {len(s0)}")
print(f"- Split array:\n {s0}" )

- Type of the split array: <class 'list'>
- Shape of the split array: 5
- Split array:
 [array([[[-1.18757848, -0.79118793,  0.70243022,  1.99311768],
        [-0.77927955, -0.86832198,  0.98496926, -1.50330293],
        [ 0.97028981,  1.28465103,  1.56144177, -0.17915802]]]), array([[[ 1.06989763, -0.43641578,  0.2285948 , -0.59473993],
        [ 1.14627666,  0.52497164,  0.59748224, -1.15890587],
        [ 0.49122778,  0.86491478,  0.27532992, -1.02358379]]]), array([[[-0.15590875,  0.80055912,  0.52980972,  0.59903495],
        [ 0.60222633, -0.30666414,  0.50599198,  1.06283019],
        [-0.38799057, -0.25074206,  0.00184736,  0.6544357 ]]]), array([[[-1.49093206,  0.95249384,  0.02130229,  2.02072982],
        [ 1.29165081,  0.87072789, -2.19943252,  0.1080727 ],
        [-0.22857142,  0.12997393,  0.82327618,  0.08925894]]]), array([[[-0.07663753, -0.65946211, -0.69698589, -1.75645967],
        [-0.22934715, -0.24436456,  0.46540788, -0.01620507],
        [-0.1250468 , -0.346722

The result is a list of five 3 by 4 arrays

In [58]:
# Split the array along the second axis
s1 = np.split(before_split, 3, axis=1)
print(f"- Type of the split array: {type(s1)}")
print(f"- Shape of the split array: {len(s1)}")
print(f"- Split array:\n {s1}" )

- Type of the split array: <class 'list'>
- Shape of the split array: 3
- Split array:
 [array([[[-1.18757848, -0.79118793,  0.70243022,  1.99311768]],

       [[ 1.06989763, -0.43641578,  0.2285948 , -0.59473993]],

       [[-0.15590875,  0.80055912,  0.52980972,  0.59903495]],

       [[-1.49093206,  0.95249384,  0.02130229,  2.02072982]],

       [[-0.07663753, -0.65946211, -0.69698589, -1.75645967]]]), array([[[-0.77927955, -0.86832198,  0.98496926, -1.50330293]],

       [[ 1.14627666,  0.52497164,  0.59748224, -1.15890587]],

       [[ 0.60222633, -0.30666414,  0.50599198,  1.06283019]],

       [[ 1.29165081,  0.87072789, -2.19943252,  0.1080727 ]],

       [[-0.22934715, -0.24436456,  0.46540788, -0.01620507]]]), array([[[ 0.97028981,  1.28465103,  1.56144177, -0.17915802]],

       [[ 0.49122778,  0.86491478,  0.27532992, -1.02358379]],

       [[-0.38799057, -0.25074206,  0.00184736,  0.6544357 ]],

       [[-0.22857142,  0.12997393,  0.82327618,  0.08925894]],

       [[-0.1

The result is a list of three 5 by 4 arrays

In [67]:
# Split the array along the third axis
s2 = np.split(before_split, 4, axis=2)
print(f"- Type of the split array: {type(s2)}")
print(f"- Shape of the split array: {len(s2)}")
print(f"- Split array:\n {s2}")

- Type of the split array: <class 'list'>
- Shape of the split array: 4
- Split array:
 [array([[[-1.18757848],
        [-0.77927955],
        [ 0.97028981]],

       [[ 1.06989763],
        [ 1.14627666],
        [ 0.49122778]],

       [[-0.15590875],
        [ 0.60222633],
        [-0.38799057]],

       [[-1.49093206],
        [ 1.29165081],
        [-0.22857142]],

       [[-0.07663753],
        [-0.22934715],
        [-0.1250468 ]]]), array([[[-0.79118793],
        [-0.86832198],
        [ 1.28465103]],

       [[-0.43641578],
        [ 0.52497164],
        [ 0.86491478]],

       [[ 0.80055912],
        [-0.30666414],
        [-0.25074206]],

       [[ 0.95249384],
        [ 0.87072789],
        [ 0.12997393]],

       [[-0.65946211],
        [-0.24436456],
        [-0.34672226]]]), array([[[  7.02430216e-01],
        [  9.84969259e-01],
        [  1.56144177e+00]],

       [[  2.28594800e-01],
        [  5.97482241e-01],
        [  2.75329919e-01]],

       [[  5.29809717e-01],

The result is a list of five 3 by 4 arrays

In [61]:
before_split

array([[[ -1.18757848e+00,  -7.91187933e-01,   7.02430216e-01,
           1.99311768e+00],
        [ -7.79279553e-01,  -8.68321983e-01,   9.84969259e-01,
          -1.50330293e+00],
        [  9.70289813e-01,   1.28465103e+00,   1.56144177e+00,
          -1.79158023e-01]],

       [[  1.06989763e+00,  -4.36415783e-01,   2.28594800e-01,
          -5.94739925e-01],
        [  1.14627666e+00,   5.24971640e-01,   5.97482241e-01,
          -1.15890587e+00],
        [  4.91227783e-01,   8.64914780e-01,   2.75329919e-01,
          -1.02358379e+00]],

       [[ -1.55908750e-01,   8.00559122e-01,   5.29809717e-01,
           5.99034948e-01],
        [  6.02226333e-01,  -3.06664144e-01,   5.05991975e-01,
           1.06283019e+00],
        [ -3.87990567e-01,  -2.50742057e-01,   1.84735730e-03,
           6.54435698e-01]],

       [[ -1.49093206e+00,   9.52493841e-01,   2.13022915e-02,
           2.02072982e+00],
        [  1.29165081e+00,   8.70727890e-01,  -2.19943252e+00,
           1.08072701

### 10. Array Shape Manipulation
#### Reshape
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html#numpy.reshape

NOTE: This will be a new view object if possible; otherwise, it will be a copy.

In [64]:
# Array to start with
my_start_array = np.array(np.arange(24))
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [65]:
# Reshape array
my_3_8_array = my_start_array.reshape((3,8))
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [69]:
# Changing an element in the reshaped array also change the element in the original array!!!
my_3_8_array[0,0] = 1234
print(f"- The reshaped array:\n {my_3_8_array}" )
print(f"- The original array:\n {my_start_array}" )

- The reshaped array:
 [[1234    1    2    3    4    5    6    7]
 [   8    9   10   11   12   13   14   15]
 [  16   17   18   19   20   21   22   23]]
- The original array:
 [1234    1    2    3    4    5    6    7    8    9   10   11   12   13   14
   15   16   17   18   19   20   21   22   23]


#### Flatten an array

In [70]:
# Flatten my_3_8_array
my_ravel_array = my_3_8_array.ravel()
my_ravel_array

array([1234,    1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
         11,   12,   13,   14,   15,   16,   17,   18,   19,   20,   21,
         22,   23])

In [71]:
# Loop through the elements of an array
for n in my_3_8_array.flat:
    print(n)

1234
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


### 11. Rearranging Array Elements

In [2]:
import numpy as np

# Create arrays to use
my_start_array = np.array(np.arange(24))
my_3_8_array = my_start_array.reshape((3,8))
my_2_3_4_array = my_3_8_array.reshape((2,3,4))

#### fliplr "flip left right"
Flip array in the left/right direction. More information can be found on: http://docs.scipy.org/doc/numpy/reference/generated/numpy.fliplr.html#numpy.fliplr

In [3]:
# For two-dimensional arrays
print(f"- The original array:\n {my_3_8_array}" )
print(f"- The flipped array:\n {np.fliplr(my_3_8_array)}" )

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The reshaped array:
 [[ 7  6  5  4  3  2  1  0]
 [15 14 13 12 11 10  9  8]
 [23 22 21 20 19 18 17 16]]


The order of the columns have been reversed

In [6]:
# For three-dimensional arrays
print(f"- The original array:\n {my_2_3_4_array}" )
print(f"- The flipped array:\n {np.fliplr(my_2_3_4_array)}" )

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The flipped array:
 [[[ 8  9 10 11]
  [ 4  5  6  7]
  [ 0  1  2  3]]

 [[20 21 22 23]
  [16 17 18 19]
  [12 13 14 15]]]


The fliplr() method acts on the last dimension

#### flipud "flip up down"
Flip array in the up/down direction. More information can be found on: http://docs.scipy.org/doc/numpy/reference/generated/numpy.flipud.html#numpy.flipud

In [5]:
# For two-dimensional arrays
print(f"- The original array:\n {my_3_8_array}" )
print(f"- The flipped array:\n {np.flipud(my_3_8_array)}" )

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The flipped array:
 [[16 17 18 19 20 21 22 23]
 [ 8  9 10 11 12 13 14 15]
 [ 0  1  2  3  4  5  6  7]]


The order of the rows have been reversed

In [8]:
# For three-dimensional arrays
print(f"- The original array:\n {my_2_3_4_array}" )
print(f"- The flipped array:\n {np.flipud(my_2_3_4_array)}" )

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The flipped array:
 [[[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]

 [[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]]


The order of the sub-arrays has been reversed

#### Rolling an array
More information can be found on: http://docs.scipy.org/doc/numpy/reference/generated/numpy.roll.html#numpy.roll

In [14]:
# For a one dimensional array
print(f"- The original array:\n {my_start_array}" )
print(f"- The rolled array:\n {np.roll(my_start_array, shift = 5)}" )

- The original array:
 [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]
- The rolled array:
 [19 20 21 22 23  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18]


The last five elements are now the first five elements of the array

In [13]:
# Now we can also roll the first elements instead of the last element by using a negative value
# for the shift parameter
np.roll(my_start_array, shift = -5)

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23,  0,  1,  2,  3,  4])

In [15]:
# For a three-dimensional array
print(f"- The original array:\n {my_2_3_4_array}" )
print(f"- The rolled array:\n {np.roll(my_2_3_4_array, shift = 2)}" )

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The rolled array:
 [[[22 23  0  1]
  [ 2  3  4  5]
  [ 6  7  8  9]]

 [[10 11 12 13]
  [14 15 16 17]
  [18 19 20 21]]]


In [16]:
# To roll the first elements
np.roll(my_2_3_4_array, -2)

array([[[ 2,  3,  4,  5],
        [ 6,  7,  8,  9],
        [10, 11, 12, 13]],

       [[14, 15, 16, 17],
        [18, 19, 20, 21],
        [22, 23,  0,  1]]])

#### Rotating an array
More information can be found on: http://docs.scipy.org/doc/numpy/reference/generated/numpy.roll.html#numpy.roll

In [17]:
# Positive rotation for a one dimensional array 
print(f"- The original array:\n {my_3_8_array}" )
print(f"- The rolled array:\n {np.rot90(my_3_8_array)}" )

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The rolled array:
 [[ 7 15 23]
 [ 6 14 22]
 [ 5 13 21]
 [ 4 12 20]
 [ 3 11 19]
 [ 2 10 18]
 [ 1  9 17]
 [ 0  8 16]]


We get a 8 by 3 array with the last column becoming the first row

In [18]:
# Negative rotation for a one dimensional array 
print(f"- The original array:\n {my_3_8_array}" )
print(f"- The rolled array:\n {np.rot90(my_3_8_array, k=-1)}" )

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The rolled array:
 [[16  8  0]
 [17  9  1]
 [18 10  2]
 [19 11  3]
 [20 12  4]
 [21 13  5]
 [22 14  6]
 [23 15  7]]


We get a 8 by 3 array with the last row becoming the first column

### 12. Transpose-like Operations
- transpose: use np.transpose to permute all the axes at once.
- swapaxes: use np.swapaxes to swap any two axes.
- rollaxes: use np.rollaxis to "rotate" the axes.

In [19]:
import numpy as np

# Arrays to use
my_start_array = np.array(np.arange(24))
my_3_8_array = my_start_array.reshape((3,8))
my_2_3_4_array = my_3_8_array.reshape((2,3,4))

#### Transpose
Transpose reverse the dimension of an array, except if a specific re-arrangement of the dimensions is given. For more information, please see: http://docs.scipy.org/doc/numpy/reference/generated/numpy.transpose.html

In [21]:
# Transpose a two-dimensional array
print(f"- The original array:\n {my_3_8_array}")
print(f"- The transposed array:\n {np.transpose(my_3_8_array)}")

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The transposed array:
 [[ 0  8 16]
 [ 1  9 17]
 [ 2 10 18]
 [ 3 11 19]
 [ 4 12 20]
 [ 5 13 21]
 [ 6 14 22]
 [ 7 15 23]]


In [22]:
# Transpose a three-dimensional array
print(f"- The original array:\n {my_2_3_4_array}")
print(f"- The transposed array:\n {np.transpose(my_2_3_4_array)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The transposed array:
 [[[ 0 12]
  [ 4 16]
  [ 8 20]]

 [[ 1 13]
  [ 5 17]
  [ 9 21]]

 [[ 2 14]
  [ 6 18]
  [10 22]]

 [[ 3 15]
  [ 7 19]
  [11 23]]]


The dimensions have been reversed so we obtained a 4 by 3 by 2 array

In [26]:
# We can choose the permutation to use by using the axes option
print(f"- The original array:\n {my_2_3_4_array}" )
print(f"- The transposed array with axes = (0,2,1):\n {np.transpose(my_2_3_4_array, axes=(0,2,1))}")
print(f"- The transposed array with axes = (2,1,0):\n {np.transpose(my_2_3_4_array, axes=(2,1,0))}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The transposed array with axes = (0,2,1):
 [[[ 0  4  8]
  [ 1  5  9]
  [ 2  6 10]
  [ 3  7 11]]

 [[12 16 20]
  [13 17 21]
  [14 18 22]
  [15 19 23]]]
- The transposed array with axes = (2,1,0):
 [[[ 0 12]
  [ 4 16]
  [ 8 20]]

 [[ 1 13]
  [ 5 17]
  [ 9 21]]

 [[ 2 14]
  [ 6 18]
  [10 22]]

 [[ 3 15]
  [ 7 19]
  [11 23]]]


#### Swapaxes
For Numpy >= 1.10, if a is an ndarray, then a view of a is returned; otherwise a new array is created. For earlier Numpy versions a view of a is returned only if the order of the axes is changed, otherwise the input array is returned. For more information: http://docs.scipy.org/doc/numpy/reference/generated/numpy.swapaxes.html#numpy.swapaxes.

In [28]:
# Swap axes in a two-dimensional array (equivalent to a transposition)
print(f"- The original array:\n {my_3_8_array}")
print(f"- The swapped array:\n {np.swapaxes(my_3_8_array, 1, 0)}")

- The original array:
 [[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]]
- The transposed array:
 [[ 0  8 16]
 [ 1  9 17]
 [ 2 10 18]
 [ 3 11 19]
 [ 4 12 20]
 [ 5 13 21]
 [ 6 14 22]
 [ 7 15 23]]


In [30]:
# Swap axes in a three-dimensional array
print(f"- The original array:\n {my_2_3_4_array}")
print(f"- The swapped array:\n {np.swapaxes(my_2_3_4_array, 1, 0)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The swapped array:
 [[[ 0  1  2  3]
  [12 13 14 15]]

 [[ 4  5  6  7]
  [16 17 18 19]]

 [[ 8  9 10 11]
  [20 21 22 23]]]


In [31]:
# Another example with a three-dimensional array
print(f"- The original array:\n {my_2_3_4_array}")
print(f"- The swapped array:\n {np.swapaxes(my_2_3_4_array, 2, 0)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The swapped array:
 [[[ 0 12]
  [ 4 16]
  [ 8 20]]

 [[ 1 13]
  [ 5 17]
  [ 9 21]]

 [[ 2 14]
  [ 6 18]
  [10 22]]

 [[ 3 15]
  [ 7 19]
  [11 23]]]


#### Rollaxis
Roll the specified axis backwards, until it lies in a given position (rotation). For Numpy >= 1.10 a view of a is always returned. For earlier Numpy versions a view of a is returned only if the order of the axes is changed, otherwise the input array is returned. For more information: http://docs.scipy.org/doc/numpy/reference/generated/numpy.rollaxis.html#numpy.rollaxis

In [32]:
# Roll axes in a three-dimensional array
print(f"- The original array:\n {my_2_3_4_array}")
print(f"- The swapped array:\n {np.rollaxis(my_2_3_4_array, 0, 2)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The swapped array:
 [[[ 0  1  2  3]
  [12 13 14 15]]

 [[ 4  5  6  7]
  [16 17 18 19]]

 [[ 8  9 10 11]
  [20 21 22 23]]]


### 13. Tiling Arrays

In [2]:
import numpy as np

# Create array to use
my_start_array = np.array(np.arange(12))
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

#### Tile
Construct an array by repeating A the number of times given by reps. 

- If reps has length d, the result will have dimension of max(d, A.ndim).
- If A.ndim < d, A is promoted to be d-dimensional by prepending new axes. So a shape (3,) array is promoted to (1, 3) for 2-D replication, or shape (1, 1, 3) for 3-D replication. If this is not the desired behavior, promote A to d-dimensions manually before calling this function.
- If A.ndim > d, reps is promoted to A.ndim by pre-pending 1’s to it. Thus for an A of shape (2, 3, 4, 5), a reps of (2, 2) is treated as (1, 1, 2, 2). 

For more information: http://docs.scipy.org/doc/numpy/reference/generated/numpy.tile.html#numpy.tile



In [3]:
# Create an array by repeating a given array 3 times
np.tile(my_start_array, 3)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,  0,  1,  2,  3,  4,
        5,  6,  7,  8,  9, 10, 11,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
       10, 11])

In [4]:
# We can reshape the obtained array
np.tile(my_start_array, 3).reshape((3,12))

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])

In [7]:
# Let's create a two-dimensional array
my_second_array = np.array(np.arange(7))
tile_1 = np.tile(my_second_array, (3, 1))
tile_1

array([[0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6]])

In [8]:
# Tile a two-dimensional array
tile_2 = np.tile(tile_1, (2,2))
tile_2

array([[0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6]])

#### Repeat
Unlike tile, this repeats element-by-element. For more information: http://docs.scipy.org/doc/numpy/reference/generated/numpy.repeat.html#numpy.repeat

In [9]:
# Repeat a one-dimensional array
print(f"- The original array:\n {my_second_array}")
print(f"- The repeated array:\n {np.repeat(my_second_array, 3)}")

- The original array:
 [0 1 2 3 4 5 6]
- The repeated array:
 [0 0 0 1 1 1 2 2 2 3 3 3 4 4 4 5 5 5 6 6 6]


In [11]:
# Repeat a three-dimensional array along the first axis
my_repeatable_array = np.array(np.arange(24)).reshape(2,3,4)
print(f"- The original array:\n {my_repeatable_array}")
print(f"- The repeated array:\n {np.repeat(my_repeatable_array, 2, axis=0)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The repeated array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]


In [12]:
# Repeat a three-dimensional array along the second axis
print(f"- The original array:\n {my_repeatable_array}")
print(f"- The repeated array:\n {np.repeat(my_repeatable_array, 2, axis=1)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The repeated array:
 [[[ 0  1  2  3]
  [ 0  1  2  3]
  [ 4  5  6  7]
  [ 4  5  6  7]
  [ 8  9 10 11]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [12 13 14 15]
  [16 17 18 19]
  [16 17 18 19]
  [20 21 22 23]
  [20 21 22 23]]]


In [13]:
# Repeat a three-dimensional array along the third axis
print(f"- The original array:\n {my_repeatable_array}")
print(f"- The repeated array:\n {np.repeat(my_repeatable_array, 2, axis=2)}")

- The original array:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
- The repeated array:
 [[[ 0  0  1  1  2  2  3  3]
  [ 4  4  5  5  6  6  7  7]
  [ 8  8  9  9 10 10 11 11]]

 [[12 12 13 13 14 14 15 15]
  [16 16 17 17 18 18 19 19]
  [20 20 21 21 22 22 23 23]]]
